src/poetry/utils/vcs/git.py

```diff
@@ -0,0 +1,193 @@
+from __future__ import annotations
+
+import re
+import shutil
+
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from dulwich import porcelain
+from dulwich.client import get_transport_and_path
+from dulwich.config import ConfigFile
+from dulwich.config import parse_submodules
+from dulwich.errors import NotGitRepository
+from dulwich.refs import ANNOTATED_TAG_SUFFIX
+from dulwich.repo import Repo
+from poetry.core.utils.helpers import canonicalize_name
+
+
+if TYPE_CHECKING:
+    from dulwich.client import FetchPackResult
+    from dulwich.client import GitClient
+
+
+class Git:
+    @staticmethod
+    def make_repo(repo: Repo | Path | str) -> Repo:
+        if not isinstance(repo, Repo):
+            repo = Repo(repo)
+        return repo
+
+    @classmethod
+    def get_remote_url(cls, repo: Repo | Path | str, remote: str = "origin") -> str:
+        repo = cls.make_repo(repo)
+        config = repo.get_config()
+        section = (b"remote", remote.encode("utf-8"))
+
+        return config.get(section, b"url") if config.has_section(section) else ""
+
+    @classmethod
+    def get_revision(cls, repo: Repo | Path | str) -> str:
+        return cls.make_repo(repo).head().decode("utf-8")
+
+    @classmethod
+    def info(cls, repo: Repo | Path | str) -> tuple[str, str]:
+        repo = cls.make_repo(repo)
+        return cls.get_remote_url(repo), cls.get_revision(repo)
+
+    @staticmethod
+    def clean(repo: Repo | Path | str) -> None:
+        repo = str(repo.path if isinstance(repo, Repo) else repo)
+        porcelain.clean(repo=repo, target_dir=repo)
+
+    @classmethod
+    def clone_submodules(cls, repo: Repo | str) -> None:
+        repo = cls.make_repo(repo)
+        repo_root = Path(repo.path)
+        modules_config = repo_root.joinpath(".gitmodules")
+
+        if modules_config.exists():
+            config = ConfigFile.from_path(modules_config)
+
+            url: bytes
+            path: bytes
+            for path, url, _ in parse_submodules(config):
+                path_relative = Path(path.decode("utf-8"))
+                path_absolute = repo_root.joinpath(path_relative)
+
+                source_root = path_absolute.parent
+                source_root.mkdir(parents=True, exist_ok=True)
+
+                revision = repo.open_index()[path].sha.decode("utf-8")
+
+                cls.clone(
+                    url=url.decode("utf-8"),
+                    source_root=source_root,
+                    name=path_relative.name,
+                    revision=revision,
+                    clean=path_absolute.exists()
+                    and not path_absolute.joinpath(".git").is_dir(),
+                )
+
+    @classmethod
+    def clone(
+        cls,
+        url: str,
+        source_root: Path,
+        name: str | None = None,
+        branch: str | None = None,
+        tag: str | None = None,
+        revision: str | None = None,
+        clean: bool = False,
+    ) -> Repo:
```

Comment:
This method is quite large. It would be easier to read and maintain if we extract some stuff into its own methods.

Reply:
Hmm, I'll see if it makes sense to break it up.